File: app/javascript/channels/chatroom_channel.js
import consumer from "./consumer";

/**
 * Subscribe to a chatroom channel
 * @param {string} chatroomId - The ID of the chatroom to subscribe to
 * @param {string} userId - The current user's ID
 * @param {function} onReceived - Callback function when a message is received
 * @returns {object} - The subscription object
 */
export const subscribeToChatroom = (chatroomId, userId, onReceived) => {
  return consumer.subscriptions.create(
    {
      channel: "ChatroomChannel",
      chatroom_id: chatroomId,
      user_id: userId,
    },
    {
      connected() {
        console.log(`Connected to chatroom ${chatroomId}`);
      },

      disconnected() {
        console.log(`Disconnected from chatroom ${chatroomId}`);
      },

      received(data) {
        console.log("Received data:", data);
        if (onReceived) {
          onReceived(data);
        }
      },
    }
  );
};

/**
 * Unsubscribe from a chatroom channel
 * @param {object} subscription - The subscription object to unsubscribe
 */
export const unsubscribeFromChatroom = (subscription) => {
  if (subscription) {
    subscription.unsubscribe();
    console.log("Unsubscribed from chatroom");
  }
};

File: app/channels/chatroom_channel.rb
class ChatroomChannel < ApplicationCable::Channel
  def subscribed
    chatroom = Chatroom.find_by(id: params[:chatroom_id])
    
    if chatroom
      stream_from "chatroom_#{chatroom.id}"
    else
      reject
    end
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
    stop_all_streams
  end

  def receive(data)
    # Handle any incoming data from the client if needed
  end
end
